specs/agents/tracing-instrumentation-db.md

```diff
@@ -1,34 +1,178 @@
+|`_.address`|e.g. `dynamodb.us-west-2.amazonaws.com`|
+|`_.port`|e.g. `5432`|
+|`_.service.name`| `dynamodb` |
+|`_.service.type`|`db`|
+|`_.service.resource`| `dynamodb` |
+|`_.cloud.region`| e.g. `us-east-1` | The AWS region where the table is, if available. |
+  
+### AWS S3
 
-For MongoDB, the span name should be the command name in the context of its collection/database, e.g. `users.find`.
+| Field | Value / Examples | Comments |
+|-------|:---------------:|----------|
+|`name`| e.g. `S3 GetObject my-bucket`|  The span name should follow this pattern: `S3 <OperationName> <bucket-name>.` Note that the operation name is in PascalCase. |
+|`type`|`storage`|
+|`subtype`|`s3`|
+|`action`| e.g. `GetObject` | The operation name in PascalCase. |
+| __**context.db._**__  |<hr/>|<hr/>|
+|`_.instance`| e.g. `us-east-1` | The AWS region where the table is. |
+|`_.statement`| :heavy_minus_sign: |  |
+|`_.type`|`dynamodb`|
+|`_.user`| :heavy_minus_sign: |
+|`_.link`| :heavy_minus_sign: |
+|`_.rows_affected`| :heavy_minus_sign: |
+| __**context.destination._**__ |<hr/>|<hr/>|
+|`_.address`|e.g. `dynamodb.us-west-2.amazonaws.com`| Not available in some cases. Only set if the actual connection is available. |
+|`_.port`|e.g. `5432`| Not available in some cases. Only set if the actual connection is available. |
+|`_.service.name`| `s3` |
+|`_.service.type`|`storage`|
+|`_.service.resource`| e.g. `accesspoint/myendpointslashes` or  `accesspoint:myendpointcolons`| The bucket name, if available. The s3 API allows either the bucket name or an Access Point to be provided when referring to a bucket. Access Points can use either slashes or colons. When an Access Point is provided, the access point name preceded by accesspoint/ or accesspoint: should be extracted. For example, given an Access Point such as `arn:aws:s3:us-west-2:123456789012:accesspoint/myendpointslashes`, the agent extracts `accesspoint/myendpointslashes`. Given an Access Point such as `arn:aws:s3:us-west-2:123456789012:accesspoint:myendpointcolons`, the agent extracts `accesspoint:myendpointcolons`. |
+|`_.cloud.region`| e.g. `us-east-1` | The AWS region where the bucket is. |
 
-For Elasticsearch, the span name should be `Elasticsearch: <method> <path>`, e.g.
-`Elasticsearch: GET /index/_search`.
+### Elasticsearch
 
-### Database span type/subtype
+| Field | Value / Examples | Comments |
+|-------|:---------------:|----------|
+|`name`| e.g. `Elasticsearch: GET /index/_search` |  The span name should be `Elasticsearch: <method> <path>` |
+|`type`|`db`|
+|`subtype`|`elasticsearch`|
+|`action`| `request` |
+| __**context.db._**__  |<hr/>|<hr/>|
+|`_.instance`| :heavy_minus_sign: |
+|`_.statement`| e.g. <pre lang="json">{"query": {"match": {"user.id": "kimchy"}}}</pre> | For Elasticsearch search-type queries, the request body may be recorded. Alternatively, if a query is specified in HTTP query parameters, that may be used instead. If the body is gzip-encoded, the body should be decoded first.|
```

Comment:
The Node.js (and I believe Python, because I copied from there) agents use a format that can include either or both of (a) query parms (in URL query encoded form) and (b) the request body (JSON) separated by two newlines:

https://github.com/elastic/apm-agent-nodejs/blob/v3.14.0/lib/instrumentation/elasticsearch-shared.js#L21-L28
https://github.com/elastic/apm-agent-python/blob/master/elasticapm/instrumentation/packages/elasticsearch.py#L65-L78

Is that a form worth codifying?

Dale, who is using this field from Kibana APM traces for Elasticsearch performance work, brought up [this discussion ticket](https://github.com/elastic/apm-agent-nodejs/issues/2019) to reconsider this format.

In general, with Elasticsearch, you need to consider the URL path, the query params, *and* the request body to fully understand the query.

Reply:
If we are just documenting the status quo here, then perhaps saying "that may be used as well" rather than "that may be used instead" is more accurate.

Reply:
That seems to be something that's specific to the Node.js and Python agents and that it can and probably should change in the future. Personally, I think it'd make more sense to use `context.http.*` to store the URL and the query parameters than to add it to `db.statement`. Ralley could then have a condition on whether the db.statement starts with `{`.